### PR TITLE
8320112 GenShen: Improve end of process report

### PR DESCRIPTION
**Pull request by earthling-amzn**
When Shenandoah is shutdown, it logs a summary of its activity over the life of the process. This change adds percentages to the breakdown of the types and counts of cycles that executed. It also simplifies some of the logic around recording the various outcomes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8320112](https://bugs.openjdk.org/browse/JDK-8320112): GenShen: Improve end of process report (**Enhancement** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/357/head:pull/357` \
`$ git checkout pull/357`

Update a local copy of the PR: \
`$ git checkout pull/357` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/357/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 357`

View PR using the GUI difftool: \
`$ git pr show -t 357`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/357.diff">https://git.openjdk.org/shenandoah/pull/357.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/357#issuecomment-1811536050)

### CODE REVIEW

**Review by kdnilsen — COMMENTED** [review]
I haven't digested all the control paths.  Some of the changes make me a bit nervous.  When degen young "upgrades to full", we have special handling.  For example, we update the live-at-end-of-last-old-mark variable, and we reset the count of consecutive degenerated cycles.  Do these changes preserve both of those behaviors?

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 1]
```diff
@@ -811,14 +802,10 @@ bool ShenandoahControlThread::service_stw_degenerated_cycle(GCCause::Cause cause
   } else {
     assert(_degen_generation->is_young(), "Expected degenerated young cycle, if not global.");
     ShenandoahOldGeneration* old = heap->old_generation();
-    if (old->state() == ShenandoahOldGeneration::BOOTSTRAPPING && !gc.upgraded_to_full()) {
+    if (old->state() == ShenandoahOldGeneration::BOOTSTRAPPING) {
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:805]: An assert has been added in `ShenandoahFullGC` to be certain the old generation is put in the `WAITING_FOR_BOOTSTRAP` state after a Full GC>

**Comment by earthling-amzn** [comment]
We had special handling for upgraded degen cycles because the handling was in the control thread's `service_stw_degenerated_cycle` function that didn't have direct knowledge of full GCs the same way `service_stw_full_cycle` did.  With this change the special handling for upgraded cycles and the handling for regular full GCs has been consolidated in `ShenandoahFullGC`. The `live-at-end-of-last-old-mark` is updated at the end of a full GC (no change here) and at the end of old marking (no change here either). The number of consecutive degenerated cycles is reset whenever a full GC or a concurrent GC completes.  

I also discovered that upgraded degen cycles were not being counted as full GCs - which is perhaps a fix we should make upstream?

**Review by kdnilsen — APPROVED** [review]
Thanks for clarifying my confusion, and thanks for these improvements.

**Pull request closed without merging**